### PR TITLE
Way of Working

### PR DESCRIPTION
**Pull request by melroy89**
A way of working proposal written in text. See Files changes tab in this Pull request.

PLEASE, review the agreements. If you do NOT agree let us know now. If you are missing something that is not added, please let us know.

Update: This will become the final version.

### CODE REVIEW

**Review by TheVillageGuy — APPROVED** [review]
"Leave IT to the contributor"

File: CONTRIBUTING.md
Comment (e569668) [inline CONTRIBUTING.md:25]: suggestion of changing `his/her PR on their behalf` -> `a PR on their behalf`

Comment (melroy89) [reply CONTRIBUTING.md:25]: agreed

**Comment by melroy89** [comment]
> "Leave IT to the contributor"

changed the text.

**Review by e569668 — APPROVED** [review]
seems fine to me. I can imagine the edge case of PRs being left around if the contributor disappears for any reason, but I imagine these are just guidelines and maintainers can agree to take over in those situations where obvious

**Comment by melroy89** [comment]
I also added:

>  Existing GitHub Organization owners are allowed to add another GitHub users to the repository or to the GitHub organization. That being said, _be careful_ who to add and discuss on [Matrix](https://matrix.to/#/#mbin:melroy.org) first.

**Comment by asdfzdfj** [comment]
since this this is apparently inspired by zeromq's C4, it may be a good idea to also add a link to the source doc (iirc it's [this one?](https://rfc.zeromq.org/spec/42/)) as a kind of further reading, or as a context for why we have decided to do it this way.

**Comment by melroy89** [comment]
> since this this is apparently inspired by zeromq's C4, it may be a good idea to also add a link to the source doc (iirc it's [this one?](https://rfc.zeromq.org/spec/42/)) as a kind of further reading, or as a context for why we have decided to do it this way.

Agreed, you are completely correct. I got inspiration from C4 by Pieter Hintjens, causing this fork to happen. I included the C4 spec in markdown format as well, extensions or differences from the spec are documented in the guidelines.md file.

**Review by TheVillageGuy — APPROVED** [review]
16 Maintainers MAY merge incorrect patches from other Contributors with the goals of (a) ending fruitless discussions, (b) capturing toxic patches in the historical record, (c) engaging with the Contributor on improving their patch quality.

introducing incorrect patches this way would not seem a good idea to me as it means compromising the working order of the project for the sake of proving someone's point. (If I am interpreting it correctly) The code should always work (*coughs*)

**Comment by melroy89** [comment]
> 16 Maintainers MAY merge incorrect patches from other Contributors with the goals of (a) ending fruitless discussions, (b) capturing toxic patches in the historical record, (c) engaging with the Contributor on improving their patch quality.
> 
> introducing incorrect patches this way would not seem a good idea to me as it means compromising the working order of the project for the sake of proving someone's point. (If I am interpreting it correctly) The code should always work (_coughs_)

I agree, it's a very strange guideline.. Not sure what the rationale behind this one was. Also it's ambitious what they mean with "incorrect". I think the GUIDELINES.md is clear to not cause regression at least with the pipeline.

**Review by nobodyatroot — APPROVED** [review]

**Comment by melroy89** [comment]
> Maintainers MAY merge incorrect patches from other Contributors with the goals of (a) ending fruitless discussions, (b) capturing toxic patches in the historical record, (c) engaging with the Contributor on improving their patch quality.

I asked around and asked in the C4 specification matrix room. And the reason is actually simple. If you merge bad code, have a record of proof in git (pull requests aren't forever it's only a github/gitlab thing).

So the idea is if you merge bad code you have proof in the git record that there is a bad actor. You can always revert the commit again or fix it. And the record can act as a proof in case the community want to get rid of bad actors.

**Comment by TheVillageGuy** [comment]
> > Maintainers MAY merge incorrect patches from other Contributors with the goals of (a) ending fruitless discussions, (b) capturing toxic patches in the historical record, (c) engaging with the Contributor on improving their patch quality.
> 
> I asked around and asked in the C4 specification matrix room. And the reason is actually simple. If you merge bad code, have a record of proof in git (pull requests aren't forever it's only a github/gitlab thing).
> 
> So the idea is if you merge bad code you have proof in the git record that there is a bad actor. You can always revert the commit again or fix it. And the record can act as a proof in case the community want to get rid of bad actors.

Oh dear, it's filled with my rubbish already :)

**Comment by BentiGorlich** [comment]
I agree with almost everything in the PR. 
2 exceptions:
 - I think we should add the guideline to always start with a `WIP` in the title. And remove it when the PR is ready to be merged. This would remove the `don't merge other peoples PR to avoid merging jnfinished code`. I know you said you'd like to merge your own PRs yourself @melroy89 , but I think to make that the general rule doesn't make a lot of sense 😅
 - I think there should be only one document that states whats true. Not one that explains the 'vanilla' C4 approach and then the changes we made.

**Comment by melroy89** [comment]
>     * I think we should add the guideline to always start with a `WIP` in the title. And remove it when the PR is ready to be merged. This would remove the `don't merge other peoples PR to avoid merging jnfinished code`. I know you said you'd like to merge your own PRs yourself @melroy89 , but I think to make that the general rule doesn't make a lot of sense 😅

I still like to merge my own PRs once I think it can be merged, despite having an approval already. But I do not agree with "WIP" since that is from Codeberg 🤣 .. but I do agree of a Draft Pull Request, that is the name that is used on GitHub for WIP PRs. I can add this to the list. But one doesn't necessarily exclude the other, right? We can still do Draft PRs and not merge another maintainer PRs unless with permission.

>     * I think there should be only one document that states whats true. Not one that explains the 'vanilla' C4 approach and then the changes we made.

Well yes, well sort of. Ideally yes 😄 . So we either deviate from C4, and we make changes to that specification. Or we put everything in contributing? Which become way too big. At the same time you noticed we have additional idea or working together (specifically to GitHub or this project), which I would like to write down... but I don't know if we should add all that to the C4? That's why I'm a bit in doubt.
Please let me know if you have a good idea to resolve this doubt I have while keeping the guidelines we find most important digestible and readable?

**Comment by BentiGorlich** [comment]
1. Lsts say it this way: I have no problem if other people merge my PRs when they are not draft PRs anymore (didn't know what they're called here). I don't think this should be a general rule, but maybe we just need more opinions on this.
2. I'd put everything in contributing, but add a short summary witt the most important points on top. But I'd say that can be confusing too, so I am not sure...

**Comment by simonrcodrington** [comment]
From my perspective, when I complete a proposal and put it up, I'm at a point where I feel the PR resolves the associated issues. I'm happy to have them modified or even merged so long as people are happy with them. 

There's still a chance that we'll merge something in and it'll break, but I'm hoping we'll be about to catch those in the PR or a little later with a bug report.  

So long as we don't let things rot as open PRs I think we'll be good

**Comment by melroy89** [comment]
Anyhow, I extend the contributing.md with the draft PR rule, see example: https://github.com/MbinOrg/mbin/blob/9af8253ae1d2a4cd09b822105c74abe59dd189d3/CONTRIBUTING.md

**Review by nobodyatroot — DISMISSED** [review]

**Review by vintprox — DISMISSED** [review]
There are several points I'm stessing about. Hope you all review those as well.

File: C4.md
Comment (vintprox) [reply C4.md:10]: Right! Just got to know that it's going to be fused with CCONTRIBUTING.md to make a singular document.

File: C4.md
Comment (vintprox) [reply C4.md:97]: I guess, we can just peruse a few statements from C4 that make sense for this project in CONTRIBUTING.md.

**Comment by melroy89** [comment]
My plan now is to indeed merge the C4 and our guidelines together to have one coherent list of rules.

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:18]: s/how/for how/

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:22]: If I understand this correctly, the meaning of the last sentence is: "Do _NOT_ merge directly to the main branch without a PR!" Is that right?

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:30]: s/right/rights/
s/contibutor/contributor/

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:31]: s/Leave it to the contributor/Let the contributor decide/
s/to also/partly to/
s/merging/merging of/

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:34]: s/. Since we did not yet define/, as we have not yet defined the/
Also, it should be either "specs say" or "spec says"

File: CONTRIBUTING.md
Comment (garrettw) [inline CONTRIBUTING.md:37]: s/loosely is/is loosely/
s/;/,/

**Review by garrettw — COMMENTED** [review]
I haven't looked over the C4 file yet; I'll do that next.

File: C4.md
Comment (PaarthShah) [inline C4.md:67]: I'm a little unsure just by reading this: does this apply to the inclusion of new libraries not written by the contributor? Because if so, I feel like it might be a bit over-restrictive.

We *should* ensure that any libraries introduced don't have negative side effects or licensing issues, but that's a whole other topic.

File: C4.md
Comment (PaarthShah) [inline C4.md:68]:
```suggestion
6.  A patch MUST compile cleanly and pass project self-tests on at least the principal target platform before a Maintainer can merge it.
```

File: C4.md
Comment (PaarthShah) [reply C4.md:80]: What about a Maintainer who wishes to contribute? Are they expected to use their own fork? Or are they allowed to create a new branch on the main repository? Or is this addressed by point 9?

File: C4.md
Comment (PaarthShah) [reply C4.md:84]: I'd go further: discussion for a patch should only be on the PR itself.

Discussion on the *problem* can be on the PR, issue, the matrix channel, et al.

This is to anticipate scenarios where multiple patches are proposed for the same problem.

File: C4.md
Comment (PaarthShah) [inline C4.md:86]: I think there's value in putting a hard deadline to respond here, to remove ambiguity and the "Oh I thought I waited long enough" style of conflict

File: C4.md
Comment (PaarthShah) [inline C4.md:90]: The PR should contain "closing keywords" to do this automatically.

File: C4.md
Comment (PaarthShah) [inline C4.md:91]: Clarify: before *and* after it's merged?

File: C4.md
Comment (PaarthShah) [inline C4.md:92]: Again, I think it's worth defining an unambiguous amount of time, *but* also allowing Maintainers to keep an issue open longer at their discretion. Stalebots etc can be used to this end.

File: C4.md
Comment (melroy89) [reply C4.md:67]: A "patch" is normally a single problem that need to be fixed. It's not really about libraries, it's about fixing this problem using a patch. The rationale behind this argument is to solve the patch by not introducing non-trivial code from others (assuming it's not causing license issues) but if a patch is needed that is non-trivial the code _could_ only come from the contributor itself. Ideally **all code** should be trivial of course and easy to understand.

File: CONTRIBUTING.md
Comment (melroy89) [reply CONTRIBUTING.md:23]: Ah yes, I need to clear this up. Of course we can merge into the main branch, but the point was to always use PRs.

File: CONTRIBUTING.md
Comment (melroy89) [reply CONTRIBUTING.md:38]: I will merge the C4 spec with our own guidelines, meaning we have one single point of truth. So this whole sentence will be removed.

File: C4.md
Comment (melroy89) [reply C4.md:69]: I agree. My proposal is to use GitHub PRs for this, explain it in the description text. And use labels where needed.

Comment (melroy89) [reply C4.md:69]: The title of the PR is key here. since that will be merged together (squashed) when merging the PR.

File: C4.md
Comment (melroy89) [reply C4.md:97]: When I combine the two guidelines, this rule will be fully removed. Maintainers / owners are now just using topic/feature branches, which is fine IMO. Or even advised.

File: C4.md
Comment (melroy89) [reply C4.md:80]: You need to fork the repo in GitHub, if you do not have the write access in GitHUb. Then you can create a PR from it. Only github org/repo members with write access can also create git branches. This is how GitHub is working.

File: C4.md
Comment (melroy89) [reply C4.md:84]: It should be a "should" indeed. Ir should be a PR and it should be discussed ideally in Matrix, especially when it's a bug fix or any other problem you're trying to fix.

We have a dedicated PR room in Matrix for PRs.

File: C4.md
Comment (PaarthShah) [reply C4.md:67]: Makes total sense to me; I think I was just over-thinking it and assumed that "patch" was also overloaded to mean "feature implementation", but from the rest of the discussion I think it's truly more limited along the lines of a bugfix.

File: C4.md
Comment (PaarthShah) [reply C4.md:80]: Right, that makes sense mechanically, I meant from a rules/expectations perspective; should maintainers also be expected to make PRs based off of their personal forks just like a non-maintainer contributor would be forced to? Or is it acceptable for them to just make a new branch on the main repository itself and make a PR from that?

**Review by asdfzdfj — COMMENTED** [review]
already mentioned in matrix and sorry for repeating this again but I think this should also be mentioned here:

stumbled upon an interesting remarks regarding this issue/PR/C4 implementation: https://programming.dev/comment/5160514

> It’s great to see the attempt and also an example of what the C4 guidelines are made to avoid.
> 
> Notice how many comments are little nitpicks about this and that. Completely stalling the commit and getting further away from the original point of C4 which is to reduce contributor friction and avoid these kind of endless discussions on PRs.
>
> I don’t want to be too critical because some of that is a clear lack of understanding of the motivations of C4 which is explained more thoroughly in Pieter’s blog posts. You don’t want to adopt a contributor guidelines that you don’t understand of course.
> 
> IMO it’s better just to implement it as-is and start using it in practice rather than bikeshedding.

and this is presumably the blog post in question, from Pieter Hintjens: [Why Optimistic Merging Works Better](http://hintjens.com/blog:106)

-----

I have to admit that after seeing the above remarks, checking out the blog and perhaps the [talk](https://www.youtube.com/watch?v=uzxcILudFWM) mentioned in the blog, I kinda agree with the remark above and that the current discussion here is starting to go into the realm of letting perfect be the enemy of good now.

And so the following review I made is with the intention to finally get this merged and actually used in practice, the refinement that @melroy89 mentioned could come and merged later

**Review by LoveIsGrief — APPROVED** [review]
Agreeing with @asdfzdfj 

This looks good as it is.

**Comment by asdfzdfj** [comment]
forgive me but I've taken the liberty to apply wording changes that @garrettw suggested to the CONTRIBUTING.md file, please recheck it again when you have the time.
(actually I have more changes I wanted to add to the guide but don't want to overstep too much at the moment)

also I found [this document](https://zguide.zeromq.org/docs/chapter6/#The-ZeroMQ-Process-C) that goes over the C4 process spec and explain thoughts behind it. highly recommened to give it a read, I think you may find some answers regarding C4 here (at least it did for me).

File: C4.md
Comment (melroy89) [reply C4.md:10]: I will update the C4.md file with our version of the C4 spec. And... I will link from CONTRIBUTING.md to this C4 file, without any exceptions in CONTRIBUTING, just a link to the C4.md file, which has all the adaptions in line with our (Mbin) WoW.

File: C4.md
Comment (melroy89) [reply C4.md:1]: Yup. I'm busy now combining our changes into OUR version of the C4 spec.

File: C4.md
Comment (melroy89) [reply C4.md:68]: Thanks. Added

File: C4.md
Comment (melroy89) [reply C4.md:69]: This section is now extended.

File: C4.md
Comment (melroy89) [reply C4.md:80]: I updated this section now. TO make it clear how it works in GitHub and how we work.

File: C4.md
Comment (melroy89) [reply C4.md:84]: I updated this section in the new version.

File: C4.md
Comment (melroy89) [reply C4.md:86]: Updated.

File: C4.md
Comment (melroy89) [reply C4.md:90]: Correct. I extend C4 with this example.

File: C4.md
Comment (melroy89) [reply C4.md:91]: After. I updated it.

File: C4.md
Comment (melroy89) [reply C4.md:92]: I remove this sentence, since we now use GitHub Stale action.

File: C4.md
Comment (melroy89) [reply C4.md:97]: I rephrased this sentence fully.

File: CONTRIBUTING.md
Comment (melroy89) [reply CONTRIBUTING.md]: I extend the C4 spec to make this all clear.

File: CONTRIBUTING.md
Comment (melroy89) [reply CONTRIBUTING.md]: I made it explicit now in C4, around 1 or 2 years inactive.

**Comment by melroy89** [comment]
As promised I updated C4 spec with our guidelines (I merged it to a single document), that matches our values and way of working now. The CONTRIBUTING.md is already big enough. So I extend/remove/updated the C4 spec accordingly to our needs and WoW and link from the Contributing guide to the C4 spec. 

I also updated the Contributing guide with some non-technical guidelines and hopefully agreements we can all agree on.

Please, review again all. 

Sorry for the delay of this PR (at least everybody had enough time to review!). I contributed first to some other PRs, commits, features, fixes for Mbin. 
I now managed to focus my time and energy to finally finish this PR. You can imagine, this is not just a "single line of code" kind of PR ;).

**Review by nobodyatroot — APPROVED** [review]

**Review by PaarthShah — APPROVED** [review]
Overall LGTM! Just minor grammar fixes/clarifying points

File: C4.md
Comment (melroy89) [reply C4.md]: True.

File: C4.md
Comment (melroy89) [reply C4.md]: I changed it now slightly. I do mean the currently active maintainers. I think it make sense if the majority of the current active maintainers are either neutral or agrees on the removal of another maintainer, this would then be the final agreement on the decision to proceed with removal.


What do you think?

Comment (PaarthShah) [reply C4.md]: Yep that's entirely sensible to me 👌🏾

Comment (melroy89) [reply C4.md]: Thank you for becoming a sponsor! I really appreciate that!! 💜

**Review by TheVillageGuy — APPROVED** [review]

**Review by nobodyatroot — APPROVED** [review]

**Review by BentiGorlich — CHANGES_REQUESTED** [review]
Other than this small inconsistency I really like this draft, good job melroy :)

File: .github/workflows/stale.yml
Comment (melroy89) [reply .github/workflows/stale.yml:10]: No it won't. PRs will not be closed. Only issues will be closed. The write permission is required to add labels to pull requests like "stale".

Comment (BentiGorlich) [reply .github/workflows/stale.yml:10]: where is it configured then? :thinking:

Comment (melroy89) [reply .github/workflows/stale.yml:10]: https://github.com/MbinOrg/mbin/blob/main/.github/workflows/stale.yml

Comment (melroy89) [reply .github/workflows/stale.yml:10]: Ps. this was added 5 months ago already ;)

Comment (BentiGorlich) [reply .github/workflows/stale.yml:10]: Ah ok :+1:

**Review by BentiGorlich — APPROVED** [review]

**Comment by BentiGorlich** [comment]
> Other than this small inconsistency I really like this draft, good job melroy :)

Misunderstanding / Misinterpretation by me, sorry :D

**Comment by melroy89** [comment]
As you can see the PR is merged. If you still have feedback you, feel free to comment on this PR or in Matrix or create an issue or a new PR.

**Pull request merged**
